packages/shopping-cart/src/managers.ts

```diff
@@ -0,0 +1,152 @@
+import debugFactory from 'debug';
+import { convertTempResponseCartToResponseCart } from './cart-functions';
+import { getEmptyResponseCart } from './empty-carts';
+import type {
+	ShoppingCartManager,
+	ShoppingCartState,
+	ResponseCart,
+	ShoppingCartManagerState,
+	ShoppingCartManagerGetState,
+	SubscribeCallback,
+	UnsubscribeFunction,
+	SubscriptionManager,
+	AddActionPromise,
+	ShoppingCartActionCreators,
+	UseShoppingCart,
+} from './types';
+
+const debug = debugFactory( 'shopping-cart:managers' );
+
+export function getShoppingCartManagerState(
+	state: ShoppingCartState,
+	lastValidResponseCart: ResponseCart,
+	areActionsPending: boolean
+): ShoppingCartManagerState {
+	const { cacheStatus, queuedActions, couponStatus, loadingErrorType, loadingError } = state;
+	const isLoading = cacheStatus === 'fresh' || cacheStatus === 'fresh-pending';
+	const isPendingUpdate =
+		queuedActions.length > 0 || cacheStatus !== 'valid' || areActionsPending !== false;
+	const loadingErrorForManager = cacheStatus === 'error' ? loadingError : null;
+	debug(
+		'manager isLoading',
+		isLoading,
+		'isPendingUpdate',
+		isPendingUpdate,
+		'loadingError',
+		loadingErrorForManager
+	);
+
+	return {
+		isLoading,
+		loadingError: loadingErrorForManager,
+		loadingErrorType,
+		isPendingUpdate,
+		couponStatus,
+		responseCart: lastValidResponseCart,
+	};
+}
+
+export function createSubscriptionManager( cartKey: string | undefined ): SubscriptionManager {
+	let subscribedClients: SubscribeCallback[] = [];
+	const subscribe = ( callback: SubscribeCallback ): UnsubscribeFunction => {
+		debug( `adding subscriber for cartKey ${ cartKey }` );
+		subscribedClients.push( callback );
+		return () => {
+			debug( `removing subscriber for cartKey ${ cartKey }` );
+			subscribedClients = subscribedClients.filter( ( prevCallback ) => prevCallback !== callback );
+		};
+	};
+	const notifySubscribers = () => {
+		debug( `notifying ${ subscribedClients.length } subscribers for cartKey ${ cartKey }` );
+		subscribedClients.forEach( ( clientCallback ) => clientCallback() );
+	};
+
+	return { subscribe, notifySubscribers };
+}
+
+export function createLastValidResponseCartManager(
+	state: ShoppingCartState
+): {
+	getLastValidResponseCart: () => ResponseCart;
+	updateLastValidResponseCart: ( state: ShoppingCartState, areActionsPending: boolean ) => void;
+} {
+	const { responseCart: initialResponseCart } = state;
+	let lastValidResponseCart = convertTempResponseCartToResponseCart( initialResponseCart );
+
+	const updateLastValidResponseCart = (
+		updatedState: ShoppingCartState,
+		areActionsPending: boolean
+	): void => {
+		const { queuedActions, cacheStatus, responseCart: tempResponseCart } = updatedState;
+		if ( queuedActions.length === 0 && cacheStatus === 'valid' && areActionsPending === false ) {
```

Comment:
Exactly the same condition (is there a pending update?) is used by the action promises object, too. It could be extracted from both objects, making them much simpler:
```js
if ( ! isStatePendingUpdate( state, areActionsPending ) ) {
  lastValidResponseCart.update( state.responseCart );
  actionPromises.resolve( state.responseCart );
}
```